### PR TITLE
adds ranges to table locks

### LINKED ISSUE #2483: Implement Table Range locks for FaTE transactions
**Is your feature request related to a problem?
The current table locks do not allow FaTE transactions operating on different parts of a table to run concurrently.

**Describe the solution you'd like**
Allow for different FaTE transactions to be able to operate on a table concurrently when they affect different ranges

User feedback for this issue captured at https://github.com/apache/accumulo/pull/2467#discussion_r801868052

### PR DESCRIPTION
**Pull request by keith-turner**
Added ranges to table locks using the ranges from fate operations.  For write locks these ranges are widened to the nearest splits, see comments in Utils.reserveTable for details.  See comments and code in DistributedReadWriteLock.WriteLock.tryLock() and
DistributedReadWriteLock.ReadLock.tryLock() for details about how these ranges change lock behavior.  Added tests to excercise this new locking behavior. Modified the `accumulo admin fate -l` command to display lock ranges.  Also modified how this command determines if a lock is held or waiting.

fixes #2483

### CODE REVIEW

**Comment by keith-turner** [comment]
This change is important for the new automatic merge feature.  Even though merges are fast now, if a merge gets stuck waiting on a long running compaction then it will block all subsequent compactions and bulk imports from running.  Adding ranges prevents blocking for the case where the ranges do not overlap.

This PR can also close #1324 #3823

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 1]
```diff
@@ -68,19 +72,22 @@ public String toString() {
...
-    private FateLockEntry(LockType lockType, FateId fateId) {
+    private FateLockEntry(LockType lockType, FateId fateId, LockRange range) {
       this.lockType = Objects.requireNonNull(lockType);
       this.fateId = Objects.requireNonNull(fateId);
+      this.range = range;
     }
 
     private FateLockEntry(String entry) {
-      var fields = entry.split(":", 2);
+      var fields = entry.split("_", 4);
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]: I like the switch to "_" since fateId uses ":". Could maybe put this in a static var "delimiter"

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 2]
```diff
@@ -251,7 +261,7 @@ public void removeEntry(FateLockEntry data, long entry) {
   public static SortedSet<NodeName> validateAndWarn(FateLockPath path, List<String> children) {
     log.trace("validating and sorting children at path {}", path);
 
-    SortedSet<NodeName> validChildren = new TreeSet<>();
+    SortedSet<NodeName> validChildren = new TreeSet<>(Comparator.comparingLong(nn -> nn.sequence));
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java:257]: What is the benefit of using a custom comparator here instead of keeping NodeName as a comparable?

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 3]
```diff
@@ -91,42 +98,57 @@ public FateId getFateId() {
...
-      return lockType.name() + ":" + fateId.canonical();
+    public LockRange getRange() {
+      return range;
     }
 
-    @Override
-    public boolean equals(Object o) {
-      if (o == null || getClass() != o.getClass()) {
-        return false;
+    private String encode(Text row) {
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]:
```suggestion
    private String encodeRow(Text row) {
```

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 4]
```diff
@@ -91,42 +98,57 @@ public FateId getFateId() {
...
+      if (row == null) {
+        return "N";
+      } else {
+        return "P" + Base64.getEncoder().encodeToString(TextUtil.getBytes(row));
       }
+    }
 
-      FateLockEntry lockEntry = (FateLockEntry) o;
-      return lockType == lockEntry.lockType && fateId.equals(lockEntry.fateId);
+    private Text decode(String enc) {
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]:
```suggestion
    private Text decodeRow(String enc) {
```

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 5]
```diff
@@ -154,26 +176,14 @@ public static class NodeName implements Comparable<NodeName> {
...
-        return this.compareTo((NodeName) o) == 0;
-      }
-      return false;
+      throw new UnsupportedOperationException();
     }
 
     @Override
     public int hashCode() {
-      return Objects.hash(sequence, fateLockEntry.get());
+      throw new UnsupportedOperationException();
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]: why are these unsupported?

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/LockRange.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ * under the License.
+ */
+package org.apache.accumulo.core.fate.zookeeper;
+
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.hadoop.io.Text;
+
+public class LockRange {
+  final KeyExtent range;
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/LockRange.java]:
```suggestion
  private final KeyExtent range;
```
Suggesting to avoid exposing the TableId.of("0") which has no significance

File: core/src/test/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLockTest.java [hunk 7]
```diff
@@ -146,10 +149,69 @@ public void testLock() throws Exception {
...
+  private DistributedLock recover(QueueLock queueLock, FateId fateId, LockRange expectedRange,
+      LockType expectedType) {
+    var dlock = DistributedReadWriteLock.recoverLock(queueLock, fateId);
+    assertEquals(expectedRange, dlock.getRange());
+    assertEquals(expectedType, dlock.getType());
+    return dlock;
+  }
+
+  @Test
+  public void testRanges() {
```
Comment (kevinrr888) [inline core/src/test/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLockTest.java:161]: You may already have this testing so ignore if I'm missing it, but it would be good to test locking at the boundaries of the ranges. For example, using the existing (E, M]. Could test locking for that range, then try to lock for range (foo, E] which should succeed and try to lock for range (M,bar] which should also succeed.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  private static final Logger log = LoggerFactory.getLogger(ComputeBulkRange.class);
+
+  private final BulkInfo bulkInfo;
+
+  public ComputeBulkRange(TableId tableId, String sourceDir, boolean setTime) {
+    BulkInfo info = new BulkInfo();
+    info.tableId = tableId;
+    info.sourceDir = sourceDir;
+    info.setTime = setTime;
+    this.bulkInfo = info;
```
Comment (kevinrr888) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java:49]:
```suggestion
    info.tableId = tableId;
    info.sourceDir = sourceDir;
    info.setTime = setTime;
    this.bulkInfo = new BulkInfo();
```

**Review by kevinrr888 — COMMENTED** [review]
I haven't looked through everything, posting comments I have. Overall, this looks like a very nice improvement to table locking.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java (hunk 8 above)
Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java:49]: This suggestion does not work because `info` is not defined.

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java (hunk 2 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java:257]: This is the only code that needs to compare NodeName objs and the code only cares about the sequence.  Adding compareTo method to the class should compare on all fields in the class, which adds a lot of complexity to the code that is not needed.  So just went w/ a simpler comparator in the place where its needed.

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java (hunk 5 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]: I suspect these are not used.  I added new fields to the class and rather than add complexity to an untested and unused method I made them throw an exception.  It is hard to tell if these are actually used, do not have good luck looking for what uses these methods in my IDE.  So if something is using them it will eventually throw an exception. 

Not sure, but I think these method were initially implemented because the class had a compareTo method and a build plugin required these methods also.

I will leave this comment open as a reminder to remove these methods after running more test.

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java (hunk 1 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java]: updated in 2bcfc43ab4fc5770025b631047a89a1eb8dc464b

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java (hunk 8 above)
Comment (kevinrr888) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/bulkVer2/ComputeBulkRange.java:49]: ah I see now, I read all the `info.` as `this.` out of habit

File: core/src/test/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLockTest.java (hunk 7 above)
Comment (keith-turner) [reply core/src/test/java/org/apache/accumulo/core/fate/zookeeper/DistributedReadWriteLockTest.java:161]: added a test in ffdc5fb1a806c55ed1c2c3c738ecfeaae5116e35

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java (hunk 2 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java:257]: > Adding compareTo method to the class should compare on all fields in the class, which adds a lot of complexity to the code that is not needed

Good point. Was worried about possible uses of NameNode in some sorted collection, with dev not realizing that they aren't comparable. But my IDE warns when you are using a sorted collection with non-comparable elts, and it falls more on the dev to understand the class they are working with.

So, I agree and think current impl is good

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java [hunk 9]
```diff
@@ -180,8 +316,9 @@ public static Lock getTableNameLock() {
     return tableNameLock;
   }
 
-  public static Lock getReadLock(Manager env, AbstractId<?> id, FateId fateId) {
-    return Utils.getLock(env.getContext(), id, fateId, LockType.READ);
+  public static DistributedLock getReadLock(Manager env, AbstractId<?> id, FateId fateId,
+      LockRange range) {
+    return Utils.getLock(env.getContext(), id, fateId, LockType.READ, range);
```
Comment (kevinrr888) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java:321]: Do `Utils.getReadLock` and `Utils.reserveTable/Namespace(...DistributedReadWriteLock.LockType.READ...)` accomplish the same thing? Can these be combined?

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java [hunk 10]
```diff
@@ -80,17 +91,142 @@ public static <T extends AbstractId<T>> T getNextId(String name, ServerContext c
...
+       * tablets that read locks need to use. For example if a write locks is obtained on (a1,c2]
+       * and it widens to (null,e] then a read lock on (d1,d9] would overlap with the widened range.
+       *
+       * Mostly widening for write locks is nice because fate operations that get read locks are
+       * probably more frequent. So the work of widening is not done on most fate operations. Also
+       * widening an infinite range is a quick operation, so create/delete table will not be slowed
+       * down by widening.
+       *
+       * Widening is done for compactions because those operations widen their range.
+       */
```
Comment (kevinrr888) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java:177]: Very good explanation. I wonder if there's a better/more visible spot for this. This spot seems fine though

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java [hunk 11]
```diff
@@ -80,17 +91,142 @@ public static <T extends AbstractId<T>> T getNextId(String name, ServerContext c
...
+      // Not expecting the lock range on a read lock to change.
+      Preconditions.checkState(widenedRange.equals(lock.getRange()));
+    }
+
+    if (lock.tryLock()) {
+      if (lockType == LockType.WRITE) {
+        // Now that table lock is acquired see if the range still widens to the same thing. If not
+        // it means the table splits changed so release the lock and try again later. Once the lock
+        // is acquired assuming the table splits in this range can not change, so this recheck is
+        // done after getting the lock.
```
Comment (kevinrr888) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java]:
```suggestion
        // it means the table splits changed so release the lock and try again later. The table splits in this range can not change once the lock is acquired, so this recheck is done after getting the lock.
```
just a bit more clear

File: core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java [hunk 12]
```diff
@@ -417,9 +430,10 @@ public void print(Map<FateInstanceType,ReadOnlyFateStore<T>> readOnlyFateStores,
 
     for (TransactionStatus txStatus : fateStatus.getTransactions()) {
       fmt.format(
-          "%-15s fateId: %s  status: %-18s locked: %-15s locking: %-15s op: %-15s created: %s%n",
+          "%-15s fateId: %s  status: %-18s locked: %-15s locking: %-15s op: %-15s created: %s lock range: %s%n",
           txStatus.getFateOp(), txStatus.getFateId(), txStatus.getStatus(), txStatus.getHeldLocks(),
-          txStatus.getWaitingLocks(), txStatus.getTop(), txStatus.getTimeCreatedFormatted());
+          txStatus.getWaitingLocks(), txStatus.getTop(), txStatus.getTimeCreatedFormatted(),
+          txStatus.getLockRange());
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java:436]: would be good to confirm if any of the fate command descriptions should be updated with this change.

File: core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java [hunk 13]
```diff
@@ -282,36 +293,42 @@ private void findLocks(ZooSession zk, final ServiceLockPath lockPath,
...
 
-            Map<FateId,List<String>> locks;
+            Map<FateId,List<String>> locks = heldLocks;
 
-            if (pos == 0) {
-              locks = heldLocks;
-            } else if (lockType == LockType.READ && !sawWriteLock) {
-              locks = heldLocks;
+            if (lockType == LockType.WRITE) {
+              for (var prev : Lists.reverse(previous)) {
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java:309]: `Lists.reverse()` might add a small amount of overhead. Its definitely more simple and readable but indexed iteration might be slightly more performant.

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java [hunk 14]
```diff
@@ -251,7 +264,7 @@ public void removeEntry(FateLockEntry data, long entry) {
   public static SortedSet<NodeName> validateAndWarn(FateLockPath path, List<String> children) {
     log.trace("validating and sorting children at path {}", path);
 
-    SortedSet<NodeName> validChildren = new TreeSet<>();
+    SortedSet<NodeName> validChildren = new TreeSet<>(Comparator.comparingLong(nn -> nn.sequence));
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java:257]: Should we refactor this comparator into a constant for reuse?

**Comment by keith-turner** [comment]
I was running bulk random walk test against this branch and found #5833.  The test is a good test for this change because it does lots of concurrent table operations with different ranges.

**Comment by keith-turner** [comment]
>  I didn't follow all of the bulk import changes

The cause of the bulk changes was the code was computing the range after it got the lock.  Needed to reorg the code to compute the range prior to getting the lock.

File: core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java (hunk 12 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java:436]: I did not see any changes that were needed.

File: core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java (hunk 13 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/AdminUtil.java:309]: I think this code is ok because its not called too frequently.

File: core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java (hunk 14 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/fate/zookeeper/FateLock.java:257]: I think this is ok because it not called too frequently.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java (hunk 9 above)
Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/Utils.java:321]: The can be used to get the lock if it exists w/o creating it in ZK if it does not.
